crux-calcite/src/crux/calcite.clj

```diff
@@ -0,0 +1,180 @@
+          {:keys [find where]} query]
+      {:find (if (seq projects) (mapv find projects) find)
+       :where (vec
+               (concat
+                (mapcat (partial ->crux-where-clauses schema) filters)
+                where))
+       :args [{:like #(org.apache.calcite.runtime.SqlFunctions/like %1 %2)}]})
+    (catch Throwable e
+      (log/error e)
+      (throw e))))
+
+(def ^:private column-types->sql-types {:varchar SqlTypeName/VARCHAR
+                                        :keyword SqlTypeName/VARCHAR
+                                        :integer SqlTypeName/INTEGER
+                                        :long SqlTypeName/BIGINT
+                                        :boolean SqlTypeName/BOOLEAN
+                                        :double SqlTypeName/DOUBLE
+                                        :datetime SqlTypeName/DATE})
+
+(defn- ^java.util.List perform-query [node q]
+  (->> (crux/q (crux/db node) q)
+       (mapv to-array)))
+
+(defn- ^String ->column-name [c]
+  (string/replace (string/upper-case (str c)) #"^\?" ""))
+
+(defn- make-table [node {:keys [:crux.sql.table/query :crux.sql.table/columns] :as table-schema}]
+  (proxy
+      [org.apache.calcite.schema.impl.AbstractTable
+       org.apache.calcite.schema.ProjectableFilterableTable]
+      []
+      (getRowType [^RelDataTypeFactory type-factory]
+        (let [field-info  (RelDataTypeFactory$Builder. type-factory)]
+          (doseq [c (:find query)]
+            (let  [col-name (->column-name c)
+                   col-type ^SqlTypeName (column-types->sql-types (columns c))]
+              (assert col-type (str "Unrecognized column: " c))
+              (log/debug "Adding column" col-name col-type)
+              (doto field-info
+                (.add col-name col-type)
+                (.nullable true))))
+          (.build field-info)))
+      (scan [root filters projects]
+        (org.apache.calcite.linq4j.Linq4j/asEnumerable
+         (perform-query node (doto (->crux-query table-schema filters projects) log/debug))))))
+
+(s/def :crux.sql.table/name string?)
+(s/def :crux.sql.table/columns (s/map-of symbol? column-types->sql-types))
+(s/def ::table (s/keys :req [:crux.db/id :crux.sql.table/name :crux.sql.table/columns]))
+
+(defn- lookup-schema [node]
+  (let [db (crux/db node)]
+    (->> (crux/q db '{:find [e]
+                      :where [[e :crux.sql.table/name]]})
+         (map (comp (partial crux/entity db) first)))))
+
+(defn create-schema [parent-schema name operands]
+  (let [node (@!crux-nodes (get operands "CRUX_NODE"))]
+    (assert node)
+    (proxy [org.apache.calcite.schema.impl.AbstractSchema] []
+      (getTableMap []
+        (into {}
+              (for [table-schema (lookup-schema node)]
+                (do (s/valid? ::table table-schema)
+                    [(string/upper-case (:crux.sql.table/name table-schema)) (make-table node table-schema)])))))))
+
+(def ^:private model {:version "1.0",
```

Comment:
I know it's 'how we do it elsewhere in Crux', but it always reads weird to me when multi-line defs are started on the same line - we'd never dream of doing this for a `defn`, and it hides arguably the most important part of the def (`model`) in the middle of the line, which makes the file harder to scan. one for the conventions chat :)